[Tutorial] OpenVINOQuantizer

daniil-lyakhov opened the pull request:
Fixes #ISSUE_NUMBER

## Description
<!--- Describe your changes in detail -->

## Checklist
<!--- Make sure to add `x` to all items in the following checklist: -->
- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

daniil-lyakhov committed:
WIP

daniil-lyakhov committed:
OpenVINOQuantizer

ynimmaga commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+
+The pytorch 2 export quantization flow uses the torch.export to capture the model into a graph and performs quantization transformations on top of the ATen graph.
+This approach is expected to have significantly higher model coverage, better programmability, and a simplified UX.
+OpenVINO is the new backend that compiles the FX Graph generated by TorchDynamo into an optimized OpenVINO model.
```
```suggestion
OpenVINO backend compiles the FX Graph generated by TorchDynamo into an optimized OpenVINO model.
```

ynimmaga commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+- Step 1: Install OpenVINO and NNCF.
+- Step 2: Capture the FX Graph from the eager Model based on the `torch export mechanism <https://pytorch.org/docs/main/export.html>`_.
+- Step 3: Apply the Quantization flow based on the captured FX Graph.
+- Step 4: Lower the quantized model into OpenVINO representation with the API ``torch.compile``.
```
```suggestion
- Step 4: Lower the quantized model into OpenVINO representation with the API `torch.compile <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`_.
```

ynimmaga commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+
+    # Create the data, using the dummy data here as an example
+    traced_bs = 50
+    x = torch.randn(traced_bs, 3, 224, 224).contiguous(memory_format=torch.channels_last)
```
why do we need the memory format to be channels_last?

daniil-lyakhov replied:
This is a copypast from the original tutorial, removed, thansk!

ynimmaga commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+    example_inputs = (x,)
+
+    # Capture the FX Graph to be quantized
+    with torch.no_grad(), disable_patching():
```
is disable_patching() needed both during export and inference with torch.compile?

daniil-lyakhov replied:
Unfortunately yes: export will fail with an error and performance of the compiled model will be ruined without it

ynimmaga commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+------------
+
+With this tutorial, we introduce how to use torch.compile with the OpenVINO backend and the OpenVINO quantizer.
+For further information, please visit `OpenVINO deploymet via torch.compile documentation <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`_.
```
```suggestion
For further information, please visit `OpenVINO deployment via torch.compile documentation <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`_.
```

alexsu52 commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+
+.. code-block:: python
+
+    quantizer = OpenVINOQuantizer()
```
```suggestion
    from nncf.experimental.torch.fx.quantization.quantizer.openvino_quantizer import OpenVINOQuantizer

    quantizer = OpenVINOQuantizer()
```

alexsu52 commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+        # from input to output nodes will be excluded from the quantization process.
+        subgraph = nncf.Subgraph(inputs=['layer_1', 'layer_2'], outputs=['layer_3'])
+        OpenVINOQuantizer(ignored_scope=nncf.IgnoredScope(subgraphs=[subgraph]))
+
```
Where can I find more information about OpenVINOQuantizer parameters?

daniil-lyakhov replied:
I've added a link to nncf API docs, which should be updated with this PR: https://github.com/openvinotoolkit/nncf/pull/3277

alexsu52 commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+Conclusion
+------------
+
+With this tutorial, we introduce how to use torch.compile with the OpenVINO backend and the OpenVINO quantizer.
```
I would suggest to add somethink like that:
For more information about NNCF and NNCF Quantization Flow for PyTorch models, please visit

daniil-lyakhov replied:
Done, please check

daniil-lyakhov commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+    from torch.ao.quantization.quantize_pt2e import prepare_pt2e
+    from torch.ao.quantization.quantizer.openvino_quantizer import OpenVINOQuantizer
+
+    import nncf
+    from nncf.torch import disable_patching
```
Unfortunately, that does not work. We can do `import nncf.torch` and then do `nncf.torch.disable_patching`

daniil-lyakhov replied:
`import nncf.torch` is introduced, please check

daniil-lyakhov committed:
Apply suggestions from code review

Co-authored-by: Alexander Suslov <alexander.suslov@intel.com>
Co-authored-by: Yamini Nimmagadda <yamini.nimmagadda@intel.com>

daniil-lyakhov commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,211 @@
+===========================================================================
+
+**Author**: dlyakhov, asuslov, aamir, # TODO: add required authors
+
```
Done

daniil-lyakhov committed:
Comments

daniil-lyakhov committed:
NNCF API docs

AlexKoff88 commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,245 @@
+
+Prerequisites
+--------------
+- `PyTorch 2 Export Post Training Quantization <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html>`_
```
```suggestion
- `PyTorch 2 Export Post-Training Quantization <https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html>`_
```

AlexKoff88 commented on prototype_source/openvino_quantizer.rst:
```diff
@@ -0,0 +1,245 @@
+Introduction
+--------------
+
+This tutorial demonstrates how to use `OpenVINOQuantizer` from `Neural Network Compression Framework (NNCF) <https://github.com/openvinotoolkit/nncf/tree/develop>`_ in PyTorch 2 Export Quantization flow to generate a quantized model customized for the `OpenVINO torch.compile backend <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`_ and explains how to lower the quantized mode ...
```
It would be more attractive if to give the user an idea why it may need to use OpenVINOQuantizer (e.g. it is more accurate, performant, etc.)

daniil-lyakhov replied:
Make sense! Description of advantages of OpenVINOQuantizer was added

daniil-lyakhov committed:
Comments

daniil-lyakhov committed:
fold_quantize=False

daniil-lyakhov committed:
Update prototype_source/openvino_quantizer.rst

Co-authored-by: Alexander Suslov <alexander.suslov@intel.com>

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer

daniil-lyakhov committed:
Spelling / comments

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer

daniil-lyakhov committed:
Apply suggestions from code review

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

daniil-lyakhov committed:
prototype_index.rst is updated

daniil-lyakhov committed:
Merge remote-tracking branch 'origin/main' into dl/fx/openvino_quantizer

daniil-lyakhov committed:
Update prototype_source/openvino_quantizer.rst

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

sekyondaMeta committed:
Remove Docs Survey Banner (#3340)

Removing the docs survey banner

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer

ZhaoqiongZ committed:
Fix code snippet format issue in inductor_windows (#3339)

Fix code snippet format issue in inductor_windows
---------

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

svekars committed:
Add a note that foreach feature is a prototype (#3341)

* Add a note that foreach feature is a prototype

AlannaBurke committed:
Updating tutorials for 2.7. (#3338)

Update the What's New section.

---------

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer

punkeel committed:
Adjust torch.compile() best practices (#3336)

* Adjust torch.compile() best practices

1. Add best practice to prefer `mod.compile` over `torch.compile(mod)`, which avoids `_orig_` naming problems.
Repro steps:
- opt_mod = torch.compile(mod)
- train opt_mod
- save checkpoint
In another script, potentially on a machine that does NOT support `torch.compile`: load checkpoint.
This fails with an error, because the checkpoint on `opt_mod` got its params renamed by `torch.compile`:
```
RuntimeError: Error(s) in loading state_dict for VQVAE:
	Missing key(s) in state_dict: "embedding.weight", "encoder.encoder.net.0.weight", "encoder.encoder.net.0.bias", ...
	Unexpected key(s) in state_dict: "_orig_mod.embedding.weight", "_orig_mod.encoder.encoder.net.0.weight", "_orig_mod.encoder.encoder.net.0.bias", ...
```
- Add best practice to use, or at least try, `fullgraph=True`. This doesn't always work, but we should encourage it.

---------

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

ZhaoqiongZ committed:
fix index format (#3343)

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

partev committed:
fix a typo in optimization_tutorial.py (#3333)

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

partev committed:
fix a typo in zeroing_out_gradients.py (#3337)

Co-authored-by: Svetlana Karslioglu <svekars@meta.com>

svekars committed:
Merge branch 'main' into dl/fx/openvino_quantizer
